File: 06-linked_list/00025-Reverse_Nodes_In_K_Group/00025-reverse_nodes_in_k_group.cpp
// 25: Reverse Nodes in K-Group
// https://leetcode.com/problems/reverse-nodes-in-k-group/


#include <iostream>
#include <vector>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

void llInput(ListNode** head, vector<int> inputs) {
    for (int i=inputs.size()-1; i>=0; i--) {
        ListNode* node = new ListNode(inputs[i], *head);
        *head = node;
    }
}

void llOutput(ListNode* head) {
    while (head != NULL) {
        cout << head->val << " -> ";
        head = head->next;
    }
    cout << "NULL" << endl;
}



class Solution {
public:
    // SOLUTION
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* dummy = new ListNode();
        dummy->next = head;

        ListNode* previous = dummy;
        ListNode* current = dummy->next;
        ListNode* t = NULL;

        int count = k;

        while (current != NULL) {
            if (count > 1) {
                t = previous->next;
                previous->next = current->next;
                current->next = current->next->next;
                previous->next->next = t;
                count--;
            } else {
                previous = current;
                current = current->next;
                count = k;

                ListNode* end = current;
                for (int i = 0; i < k; i++) {
                    if (end == NULL) return dummy->next;
                    end = end->next;
                }
            }
        }

        return dummy->next;
    }
};



int main() {
    Solution o;
    ListNode* ll = NULL;

    // INPUT
    vector<int> li = {1,2,3,4,5};
    int k = 3;
    llInput(&ll, li);

    // OUTPUT
    auto result = o.reverseKGroup(ll, k);
    llOutput(result);

    return 0;
}

File: 06-linked_list/00025-Reverse_Nodes_In_K_Group/00025-reverse_nodes_in_k_group.py
# 25: Reverse Nodes in K-Group
# https://leetcode.com/problems/reverse-nodes-in-k-group/


class ListNode:
    def __init__(self, x: int):
        self.val = x
        self.next = None

class LinkedList:
    def __init__(self) -> None:
        self.head = None

    def Input(self, inputs: list[int]):
        for i in inputs:
            node = ListNode(i)
            if self.head:
                tail = self.head
                while tail.next != None: tail = tail.next
                tail.next = node
            else:
                self.head = node
    
    def Output(self, head: ListNode):
        while (head != None):
            print(head.val, "-> ", end='')
            head = head.next
        print("NULL")


class Solution:
    # SOLUTION
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        dummy = ListNode(None)
        dummy.next = head

        previous = dummy
        current = dummy.next
        t = ListNode(None)

        count = k

        while (current != None):
            if (count > 1):
                t = previous.next
                previous.next = current.next
                current.next = current.next.next
                previous.next.next = t
                count-=1
            else:
                previous = current
                current = current.next
                count = k

                end = current
                for i in range(k):
                    if (end == None): return dummy.next
                    end = end.next

        return dummy.next


if __name__ == "__main__":
    o = Solution()
    ll = LinkedList()

    # INPUT
    li = [1,2,3,4,5]
    k = 3
    ll.Input(li)

    # OUTPUT
    result = o.reverseKGroup(ll.head, k)
    ll.Output(result)
